engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/semantic/RuntimeManagementTest.scala

```diff
@@ -208,5 +208,45 @@ class RuntimeManagementTest extends InterpreterTest {
       def all                         = 0.to(4).map(mkAccessStr) ++ List(1, 3).map(mkFreeStr)
       totalOut should contain theSameElementsAs all
     }
+
+    "Allow for multithreaded polyglot class loading" in {
+      val langCtx = interpreterContext
+        .ctx()
+        .getBindings(LanguageInfo.ID)
+        .invokeMember(MethodNames.TopScope.LEAK_CONTEXT)
+        .asHostObject[EnsoContext]()
+
+      val code =
+        """import Standard.Base.Data.Numbers
+          |polyglot java import org.enso.example.TestClass
+          |main =
+          |    instance = TestClass.new (x -> x * 2)
+          |    instance.callFunctionAndIncrement 10
+          |""".stripMargin
+
+      val main = getMain(code)
+
+      def runMain()
+        : java.util.concurrent.CompletableFuture[org.graalvm.polyglot.Value] =
+        langCtx.getThreadManager.submit(() => {
+          main.execute()
+        })
+
+      def runTest(): Unit = {
+        val futures = 0.until(parallelism).map(_ => runMain())
+        val combinedFuture = java.util.concurrent.CompletableFuture
+          .allOf(futures: _*)
+          .thenApply(_ => {
+            futures
+              .map(_.get(10, java.util.concurrent.TimeUnit.SECONDS).asInt())
+          })
+        val result =
+          combinedFuture.get(20, java.util.concurrent.TimeUnit.SECONDS)
+        result should equal(List(21, 21, 21, 21, 21))
+        futures.forall(_.isDone) shouldBe true
```

Comment:
I have also tried to simulate the problem with language initialization deadlock. Here is the basic test:
```diff

```
Copy into `X.diff` file and then:
```
$ patch -p0 <X.diff
$ JAVA_HOME=~/bin/graalvm-24/ mvn install
...
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```
e.g. it doesn't deadlock. The problem (of Enso setup) must be deeper.

Reply:
f562e15f6042446df0214e7164c5a0fa71854a97 is a nicer result of the experiments. Thanks for trying it.